### PR TITLE
oobmigrations: Startup check

### PR DESCRIPTION
**Pull request by efritz**
Update frontend startup to check out of band migration status and block startup if there are unfinished up or down migrations that may result in data loss. Fixes https://github.com/sourcegraph/sourcegraph/issues/18240. Fixes https://github.com/sourcegraph/sourcegraph/issues/20906. Fixes https://github.com/sourcegraph/sourcegraph/issues/18241.

Summary of changes (review by commit):

- DB Changes:
  - Add `first_version` field to `versions` table that is populated with the first value of `version` written to the table (via upsert or migration)
  - Add `is_enterprise` flag to `out_of_band_migrations` table
  - Split `introduced` and `deprecated` columns into `introduced_version_{major,minor}` and `deprecated_version_{major,minor}` columns
- Code changes:
  - Add utility method in `cmd/frontend/backend` package to query it new `first_version` column
  - Modify `oobmigration/Store.List` to omit enterprise-only migrations in OSS mode
  - Add `oobmigration/Runner.Validate` method that tells which migrations are breaking their guarantees based on the current instance version
  - Invoke `Validate` on frontend startup

### CODE REVIEW

**Comment by efritz** [comment]
@daxmc99 (cc @sourcegraph/distribution) We've made some progress on this PR and wanted to bring you in (re: https://github.com/sourcegraph/sourcegraph/pull/18657#discussion_r634826667). Let us know what concerns you have and where answers to those concerns should be documented for distribution/CS/administrators.

File: client/web/src/site-admin/SiteAdminMigrationsPage.tsx [hunk 1]
```diff
@@ -351,56 +350,52 @@ const MigrationNode: React.FunctionComponent<MigrationNodeProps> = ({ node, now
...
-const extract = (version: string): Version | undefined => {
-    const match = version.match(/^(\d+)\.(\d+)/)
-    if (!match) {
-        return undefined
+/** Parse the given version safely. */
+const parseVersion = (version: string): PartialVersion => {
+    try {
+        return _parseVersion(version)
+    } catch {
+        console.log('NO', version)
```
Comment (eseliger) [inline client/web/src/site-admin/SiteAdminMigrationsPage.tsx:379]:
```suggestion
```

**Review by eseliger — APPROVED** [review]

File: migrations/frontend/1528395823_feature_flags.down.sql [hunk 2]
```diff
@@ -4,5 +4,6 @@ DROP TABLE IF EXISTS feature_flag_overrides;
```
[the remark this answers is not in the document]
Comment (efritz) [reply migrations/frontend/1528395823_feature_flags.down.sql:4]: No, but it fixes down migrations and I was tired of manually fixing my schema while testing this PR.

[the remark this answers is not in the document]
Comment (eseliger) [reply migrations/frontend/1528395823_feature_flags.down.sql:4]: Ok cool, was just trying to understand how this relates.

**Comment by efritz** [comment]
Opened https://github.com/sourcegraph/sourcegraph/pull/21670 for doc changes.

**Comment by daxmc99** [comment]
Looks good! Sorry for the late review

**Pull request merged**